### PR TITLE
roachpb: move range tree protos to package storage

### PR DESCRIPTION
**Pull request by tamird**
Somewhat minor cleanup spotted while working on other stuff.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5839)

### CODE REVIEW

**Comment by petermattis** [comment]
How is moving the range tree stuff into its own package a cleanup? Are we expecting to use the range tree code outside of the `storage` package?

**Comment by tamird** [comment]
Decoupling the protos is the cleanup.

**Comment by tamird** [comment]
The range tree itself could have stayed in the storage package, I suppose. The only win is moving the protos out of metadata.proto.

Would you prefer that?

**Comment by petermattis** [comment]
Yeah, I'd prefer to keep the range tree code in the `storage` package. Moving the `RangeTree` protos into `storage/range_tree.proto` sounds fine.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEWgC4jD0FR4T2YDbD4:1104808501)*

**Comment by tamird** [comment]
Done.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEWkiy4hFaf-Pqhm651:-499146009)*

**Comment by petermattis** [comment]
LGTM, though @BramGrunier should take a look at the test changes.

---

Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/range_tree.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEWlHd2Jn6pJuzJ6mDP:-KEWlHd3slACo24IBUQT:1651617945) ([raw file](https://github.com/cockroachdb/cockroach/blob/858ced4f9dc19d622ad00c187e8e4ef09ee98457/storage/range_tree.go#L43)):*
Why the name change for this method? It still seems specific to the range tree.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEWlMiQ9cyvibPelwM1:-1459021503)*

File: storage/range_tree.go [hunk 1]
```diff
@@ -33,18 +33,18 @@ type cachedNode struct {
...
 	dirty bool
 	nodes map[string]cachedNode
 }
 
-// SetupRangeTree creates a new RangeTree. This should only be called as part
+// Setup creates a new RangeTree. This should only be called as part
 // of store.BootstrapRange.
-func SetupRangeTree(batch engine.Engine, ms *engine.MVCCStats, timestamp roachpb.Timestamp, startKey roachpb.RKey) error {
-	tree := &roachpb.RangeTree{
+func Setup(batch engine.Engine, ms *engine.MVCCStats, timestamp roachpb.Timestamp, startKey roachpb.RKey) error {
```
Comment (bdarnell) [inline storage/range_tree.go]: This rename is not appropriate without the package move.

**Comment by tamird** [comment]
Yep, @petermattis pointed that out. Fixed.

---

Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/range_tree.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEWlHd2Jn6pJuzJ6mDP:-KEWlswnCrRyW2zHzmr0:-971285104) ([raw file](https://github.com/cockroachdb/cockroach/blob/858ced4f9dc19d622ad00c187e8e4ef09ee98457/storage/range_tree.go#L43)):*
Ah, my bad - I had renamed it when this stuff was moved to another package. I've renamed it back.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEWmpYBrbaIK-NahHar:1705114517)*

**Comment by BramGruneir** [comment]
I'm about to head out on a 10 hour drive.  I won't be able to take a look until tomorrow.

---

Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEX3Nx7jEByzex47oKW:559441727)*

**Comment by BramGruneir** [comment]
Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/client_range_tree_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEX3a4dPnadrKKS2pNA:-KEX3a4dPnadrKKS2pNB:-460511688) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d3ab7af4063a2379ef4899f84bf9168c9e9e69e/storage/client_range_tree_test.go#L155)):*
childBlackCount?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839)*

**Comment by tamird** [comment]
Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/client_range_tree_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEX3a4dPnadrKKS2pNA:-KEX5IEDtjRYfa5_mUNr:-289223293) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d3ab7af4063a2379ef4899f84bf9168c9e9e69e/storage/client_range_tree_test.go#L155)):*
what's the question here? it's the count of black nodes from this node's children to their leaves.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839)*

**Comment by BramGruneir** [comment]
Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/client_range_tree_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEX3a4dPnadrKKS2pNA:-KEX9CWJu2aal1WBsv4P:869793528) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d3ab7af4063a2379ef4899f84bf9168c9e9e69e/storage/client_range_tree_test.go#L155)):*
s/childBlackCound/childBlackCount

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839)*

**Comment by petermattis** [comment]
:lgtm: We're not actually using the range-tree code in production (the range-tree is maintained, but nothing is using it). Not sure that it is was worthwhile to change the tests, but it is done. 

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/client_range_tree_test.go, line 79 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEXXFmgMN7YugGVVIzf:-KEXXFmgMN7YugGVVIzg:-2077523187) ([raw file](https://github.com/cockroachdb/cockroach/blob/40a7933f28678d6b31af571e7ef4294f59fcdefb/storage/client_range_tree_test.go#L79)):*
The comment about the node existing is no longer accurate. And this isn't the same helper as the one embedded in the range tree code (I'd just remove that sentence).

---

*[storage/client_range_tree_test.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEXYg7KelyubWg9cRVl:-KEXYg7KelyubWg9cRVm:381091850) ([raw file](https://github.com/cockroachdb/cockroach/blob/40a7933f28678d6b31af571e7ef4294f59fcdefb/storage/client_range_tree_test.go#L86)):*
`s/left/node/g`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEXY2FpZ-QSSmuMCIH3:2070117409)*

**Comment by tamird** [comment]
Most of this work was done after porting metadata.proto to proto3 caused the tests to fail. The refactoring you see here is the result of my trying to figure out what the cause of the failure was, so while it wasn't worth doing for its own sake, it is as you say, done.

The cause of the failures turned out to be the reliance of the range tree code on the distinguishability of empty keys from nil keys; proto3 encodes neither, which breaks the implementation.

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/client_range_tree_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEX3a4dPnadrKKS2pNA:-KEXxw-PO_mWm_XS08qN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d3ab7af4063a2379ef4899f84bf9168c9e9e69e/storage/client_range_tree_test.go#L155)):*
Done.

---

*[storage/client_range_tree_test.go, line 79 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEXXFmgMN7YugGVVIzf:-KEXyVZvhNi2yFxncd0V:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/40a7933f28678d6b31af571e7ef4294f59fcdefb/storage/client_range_tree_test.go#L79)):*
Done.

---

*[storage/client_range_tree_test.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEXYg7KelyubWg9cRVl:-KEXxvQ3xMSEAyRzK5F7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/40a7933f28678d6b31af571e7ef4294f59fcdefb/storage/client_range_tree_test.go#L86)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEXxwZF8mOb1AAuoqbL:527882608)*

**Comment by tamird** [comment]
Test failure is unrelated acceptance timeout.

**Comment by BramGruneir** [comment]
The proto move is good. 

Now the test changes are fine, but they were specifically designed to be a mirror of the ones in `range_tree_test.go`.  We should change both and not just the one.  But more than that, they were written to match the most common way to code up the algorithm.  http://en.literateprograms.org/Red-black_tree_(Java) is one of the many sources I used.  VerifyProperty5 does work in the bottom up way you've rewritten it, but the error messages will not as clear, since they might be in the middle or at the root of the tree.

I would suggest reverting the unneeded changes to the tests or changing the tests in `range_tree_test.go` to match.   If you stick with the new changes, isRed() needs to be removed since it is not needed without the pointers.

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/client_range_tree_test.go, line 78 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-KEb8IYYyWlpFbm6mejj:-KEb8IYZ2KoQPLCB1MKI:-1559015573) ([raw file](https://github.com/cockroachdb/cockroach/blob/19dabe06d14ecf3c1ae94e3d69ce6525de369da2/storage/client_range_tree_test.go#L78)):*
This function is useless now.  Anywhere it's used can just be `!node.Black`.  In the standard algorithm, it's required because nil nodes are considered black.  By remove the pointers, I worry that it might miss some conditions.  I'm going to go through each of the cases to make sure the conditions still hold correctly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEbFsaRRNTUppAf54Ff:-1987705686)*

**Comment by tamird** [comment]
AFAICT, the change to `VerifyProperty5` was required. I believe it was previously not verifying the invariant anywhere but the root.

I believe the new error messages are more useful since they describe the narrowest violation of the invariant.

I don't know why `isRed` "needs" to be removed - yeah, it's not really useful, but it's not incorrect. I'm happy to review a PR that changes `range_tree_test`, but I don't see a reason that the two tests need to be exact transcriptions. If that was the intent, the shared test code should have been explicit. Doing that refactoring now seems hard to justify outside of Friday time, given the liveness of this code.

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEbMyKSEJ-9Q_r5Lj0u:-1489309646)*

**Comment by BramGruneir** [comment]
I see your point on the error message.  But by checking every leaf, it means it holds for the full tree and every node implicitly.  Anyway, that change is fine.

However, on the code being similar to the one is range_tree_test, the test code specifically mentions that:
```
// These verify functions are similar to the those found in the range_tree_test
// but these use a map of nodes instead of a tree context.
```

It just seems very strange to change them here and not there.  

You started this refactor, why are you pushing the rest of it onto me?  The code was fine before and as you mentioned, it's likely to be removed.  But if you insist on changing the tests, it should be done correctly and in both contexts. 

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEbP2zYhIqPND64AazF:599955774)*

**Comment by tamird** [comment]
I went down the road of this refactor because I was investigating test failures under proto3 related to null-vs-absent; the refactor wasn't the intent, but I submitted the change because the new code was better than the old, and was already written.

I'm not pushing anything onto you. The code is fine now, too. These tests are correct. If you can help me understand why these tests need to be exactly the same as the ones in `range_tree_test`, I'll make that change, but until then, I don't see a reason to do so.

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEbVfATeU8GWLGtgIC1:921454934)*

**Comment by BramGruneir** [comment]
The reason for the mirrored code was to ensure that if the tree failed in one and not the other, it would be easier to isolate the issue.  When I first wrote the tests, I remember discovering exactly that type of issue. 

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5839#-:-KEbXBOikqmlv-4p1PYj:1649467223)*

**Pull request merged**